src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.cs

```diff
@@ -0,0 +1,1244 @@
+         * GetIndexOfFirst* and IsAscii routines
+         *
+         * Given a buffer, returns the index of the first element in the buffer which
+         * is a non-ASCII byte, or -1 if the buffer is empty or all-ASCII. The bool-
+         * returning method is a convenience shortcut to perform the same check.
+         */
+
+        /// <summary>
+        /// Finds the first non-ASCII <see cref="byte"/> within the specified buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer whose contents should be examined.</param>
+        /// <returns>The zero-based index in <paramref name="buffer"/> where the first non-ASCII
+        /// <see cref="byte"/> appears; else -1 if <paramref name="buffer"/> is empty or
+        /// contains only ASCII <see cref="byte"/>s.</returns>
+        public static unsafe int GetIndexOfFirstNonAsciiByte(ReadOnlySpan<byte> buffer)
+        {
+            fixed (byte* pBuffer = &MemoryMarshal.GetReference(buffer))
+            {
+                int result = (int)ASCIIUtility.GetIndexOfFirstNonAsciiByte(pBuffer, (uint)buffer.Length);
+                if (result == buffer.Length)
+                {
+                    result = -1; // no non-ASCII data found
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Finds the first non-ASCII <see cref="char"/> within the specified buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer whose contents should be examined.</param>
+        /// <returns>The zero-based index in <paramref name="buffer"/> where the first non-ASCII
+        /// <see cref="char"/> appears; else -1 if <paramref name="buffer"/> is empty or
+        /// contains only ASCII <see cref="char"/>s.</returns>
+        public static unsafe int GetIndexOfFirstNonAsciiChar(ReadOnlySpan<char> buffer)
+        {
+            fixed (char* pBuffer = &MemoryMarshal.GetReference(buffer))
+            {
+                int result = (int)ASCIIUtility.GetIndexOfFirstNonAsciiChar(pBuffer, (uint)buffer.Length);
+                if (result == buffer.Length)
+                {
+                    result = -1; // no non-ASCII data found
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Determines whether a buffer contains only ASCII <see cref="byte"/>s.
+        /// </summary>
+        /// <param name="value">The buffer whose contents should be examined.</param>
+        /// <returns><see langword="true"/> if <paramref name="value"/> is empty or
+        /// contains only ASCII <see cref="byte"/>s; otherwise, <see langword="false"/>.</returns>
+        /// <remarks>
+        /// This method is behaviorally equivalent to calling <see cref="GetIndexOfFirstNonAsciiByte"/>
+        /// and comparing the return value against -1.
+        /// </remarks>
+        public static bool IsAscii(ReadOnlySpan<byte> value) => GetIndexOfFirstNonAsciiByte(value) < 0;
```

Comment:
If this method is hot, this could be speeded up by not going via the index and just check if anywhere non-ASCII is present.

(I've seen your note about perf int the top comment, but still want to point out so it's not forgotten...)

Reply:
Yeah. It would require surgical changes to the workhorse method on `AsciiUtility` to do the early exit here. It's definitely feasible, but I'm not confident I could get those changes through code review. Especially since the current implementation is good enough and does only a tiny amount of extra work.

Related: `Utf8Utility`'s transcoding workhorse routines perform bounds checking on the write buffer all over the place. Those checks show up in profiles. We could eliminate those checks entirely if we knew ahead of time the write buffer was large enough to hold the worst case output.

Reply:
If we had proper templates I guess this wouldn't be quite so bad since the compiler would prune unneeded code paths. :)

We've abused generics before to achieve much the same effect. See https://source.dot.net/System.Private.CoreLib/R/7e55824695e6ac34.html.